compiler/rustc_codegen_llvm/src/context.rs

```diff
@@ -389,7 +390,19 @@ impl<'ll, 'tcx> CodegenCx<'ll, 'tcx> {
         // linker will take care of everything. Fixing this problem will likely
         // require adding a few attributes to Rust itself (feature gated at the
         // start) and then strongly recommending static linkage on Windows!
-        let use_dll_storage_attrs = tcx.sess.target.is_like_windows;
+        let use_dll_storage_attrs = tcx.sess.target.is_like_windows
+            // We don't support dynamic linking with ThinLTO so if it is enabled,
+            // we know that there is no need to mark items `dllimport`/`dllexport`.
+            // Fixes #81408 which occurs because lld does not fix up our mismatched
+            // `dllimport`/`dllexport` annotations.
+            // Note: eventually if we want to support a mixed ThinLTO & dynamic linking
+            // scenario where some crates are statically linked via ThinLTO and others
+            // are dynamically linked, then this check needs to become more precise
```

Comment:
https://github.com/rust-lang/rust/pull/101403 will enable LTO (fat and thin) for rust dynamic libraries and use on rustc itself.

Reply:
Also shouldn't dllimport be valid even for items in the same dll/exe?

Reply:
Does #101403 handle the problem described here? 

https://github.com/rust-lang/rust/blob/657f246812ab2684e3c3954b1c77f98fd59e0b21/compiler/rustc_session/src/session.rs#L1429-L1441

Reply:
> Also shouldn't dllimport be valid even for items in the same dll/exe?

I'm not sure how it could since the symbols don't line up. I also tried marking the definition with `dllexport` to try to get that to work but it doesn't seem to work either under ThinLTO.

Reply:
> Does https://github.com/rust-lang/rust/pull/101403 handle the problem described here?

That problem seems to be about linker plugin LTO, but #101403 uses rust LTO.

> I'm not sure how it could since the symbols don't line up.

It has to as in case of `extern "C" { ... }` we don't know yet if it will be a dll import or a local import. Not using dllimport would break if it is actually imported from a dll, so we have to unconditionally use it if we don't know if it is imported from a dll or not afaik.

Reply:
> > Also shouldn't dllimport be valid even for items in the same dll/exe?
> 
> I'm not sure how it could since the symbols don't line up. I also tried marking the definition with `dllexport` to try to get that to work but it doesn't seem to work either under ThinLTO.

Was it AV as well or some other problem?

Reply:
Ah, ok I see. We can move this check closer to the point where we decide if we should apply `dllimport` or not specifically for Rust items, not for `extern "C"` symbols. I also gave this patch a try rebased on top of #101403 and after enabling ThinLTO in `config.toml`, I can confirm that both link.exe and lld can link rustc correctly. So based on that, I think the scenario I was originally concerned about should work fine.

Reply:
> Was it AV as well or some other problem?

Sorry @mati865, GitHub didn't show me your comment for some reason. It was the same AV as happens currently.

Reply:
No problem.
It bothers me why it does work with `windows-gnu` with ThinLTO and LLD, I have gut feeling one of LLD's magic tricks when running in MinGW mode somehow fixes the problem. One of them could be enabled by passing `-stdcall-fixup` (depending on the context it does more than the name implies) to LLD but I guees it's not worth to be bothered by it.